app/validators/schema_validator.rb

```diff
@@ -1,35 +1,56 @@
 require 'json-schema'
 
 class SchemaValidator
-  def initialize(payload, type:, schema_name: nil, schema: nil)
-    @payload = payload
+  def initialize(type:, schema_name: nil, schema: nil)
     @type = type
     @schema = schema
     @schema_name = schema_name
   end
 
-  def validate
+  def validate(payload)
+    @payload = payload
+
     return true if schema_name_exception?
-    validate_schema
+
+    errors = JSON::Validator.fully_validate(
+      schema,
+      payload,
+      errors_as_objects: true,
+    )
+
+    return true if errors.empty?
+
+    errors = errors.map { |e| present_error e }
+
+    Airbrake.notify_or_ignore(
+      {
+        error_class: "SchemaValidationError",
+        error_message: "Error validating payload against schema"
+      },
+      parameters: {
+        errors: errors,
+        message_data: payload
+      }
+    )
+    false
   end
 
 private
 
   attr_reader :payload, :type
 
-  def validate_schema
-    JSON::Validator.validate!(schema_for_validation, payload)
-  rescue JSON::Schema::ValidationError => error
-    Airbrake.notify_or_ignore(error, parameters: {
-      explanation: "#{payload} schema validation error"
-    })
-    false
-  end
+  def present_error(error_hash)
+    # The schema key just contains an addressable, which is not informative as
+    # the schema in use should be clear from the error class and message
+    error_hash = error_hash.reject { |k, _| k == :schema }
+
+    if error_hash.has_key? :errors
+      error_hash[:errors] = Hash[
+        error_hash[:errors].map { |k, errors| [k, errors.map { |e| present_error e }] }
```

Comment:
This would probably be clearer with the outer block as a do...end:

    error_hash[:errors].map do |k, errors|
        [k, errors.map { |e| present_error(e) }]
    end